pep-0xxx.rst

```diff
@@ -0,0 +1,1250 @@
+manage TLS configuration. The goals of this class are as follows:
+
+1. To provide a method of specifying TLS configuration that avoids the risk of
+   errors in typing (this excludes the use of a simple dictionary).
+2. To provide an object that can be safely compared to other configuration
+   objects to detect changes in TLS configuration, for use with the SNI
+   callback.
+
+This class is not an ABC, primarily because it is not expected to have
+implementation-specific behaviour. The responsibility for transforming a
+``TLSConfiguration`` object into a useful set of configuration for a given TLS
+implementation belongs to the Context objects discussed below.
+
+This class has one other notable property: it is immutable. This is a desirable
+trait for a few reasons. The most important one is that it allows these objects
+to be used as dictionary keys, which is potentially extremely valuable for
+certain TLS backends and their SNI configuration. On top of this, it frees
+implementations from needing to worry about their configuration objects being
+changed under their feet, which allows them to avoid needing to carefully
+synchronize changes between their concrete data structures and the
+configuration object.
+
+This object is extendable: that is, future releases of Python may add
+configuration fields to this object as they become useful. For
+backwards-compatibility purposes, new fields are only appended to this object.
+Existing fields will never be removed, renamed, or reordered.
+
+The ``TLSConfiguration`` object would be defined by the following code::
+
+    ServerNameCallback = Callable[[TLSBufferObject, Optional[str], TLSConfiguration], Any]
+
+
+    _configuration_fields = [
+        'validate_certificates',
+        'certificate_chain',
+        'ciphers',
+        'inner_protocols',
+        'lowest_supported_version',
+        'highest_supported_version',
+        'trust_store',
+        'sni_callback',
+    ]
+
+
+    _DEFAULT_VALUE = object()
+
+
+    class TLSConfiguration(namedtuple('TLSConfiguration', _configuration_fields)):
+        """
+        An immutable TLS Configuration object. This object has the following
+        properties:
+
+        :param validate_certificates bool: Whether to validate the TLS
+            certificates. This switch operates at a very broad scope: either
+            validation is enabled, in which case all forms of validation are
+            performed including hostname validation if possible, or validation
+            is disabled, in which case no validation is performed.
+
+            Not all backends support having their certificate validation
+            disabled. If a backend does not support having their certificate
+            validation disabled, attempting to set this property to ``False``
+            will throw a ``TLSError`` when this object is passed into a
+            context object.
+
+        :param certificate_chain Tuple[Tuple[Certificate],PrivateKey]: The
```

Comment:
Oh, I understand: you want to enforce immutability and hashability. Fair enough.